[Dashboard] Fix timezone_utils to deal with daylight savings

bhmiller opened the pull request:
The `Dashboard Server Timezone` is currently incorrect in when
running in a timezone with daylight savings. Fixing that
resolution with python `tzlocal`.


## Why are these changes needed?

See https://github.com/ray-project/ray/issues/51310

## Related issue number

Closes #51310

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

My testing was just using python code through the repl.

bhmiller commented on python/ray/dashboard/client/src/common/timezone.ts:
```diff
@@ -1,794 +1,676 @@
   },
   {
-    utc: "GMT-06:00",
-    value: "America/Guadalajara",
```
Did not exist in `Intl.DateTimeFormat`

bhmiller commented on python/ray/dashboard/client/src/common/timezone.ts:
```diff
@@ -1,794 +1,676 @@
   },
   {
-    utc: "GMT+07:00",
-    value: "Asia/Hanoi",
```
Did not exist in `Intl.DateTimeFormat`

bhmiller commented on python/requirements.txt:
```diff
@@ -62,3 +62,4 @@ py-spy>=0.2.0; python_version < '3.12'
 py-spy>=0.4.0; python_version >= '3.12'
 memray; sys_platform != "win32" # memray is not supported on Windows
 pyOpenSSL
+tzlocal
```
I'm not sure how to get this available for your tests. When I ssh-ed into ray head, I saw `tzlocal`, so I _think_ it's available normally, but your tests were failing with not being able to find `tzlocal`. I could use some help pointing to how to get the tests happy as I'm otherwise just skimming through the codebase and guessing.

aslonnie commented on python/ray/dashboard/timezone_utils.py:
```diff
@@ -48,9 +10,6 @@ def get_current_timezone_info():
-        (tz for tz in timezones if tz["offset"] == current_offset),
-        {"offset": None, "value": None},
-    )
+    current_timezone = {"offset": current_offset, "value": tzlocal.get_localzone_name()}
```
why this `get_current_timezone_info` is required at the first place? for frontend ui, it should either use browser time or use utc, right? why does server side local timezone even matter?

bhmiller replied:
I did not implement this feature of yours. I'm trying to fix that the feature doesn't work. If you would rather drop the server timezone and just use the browser's time zone, that's fine by me.

aslonnie replied:
yeah, I am more of asking the component owner.

alanwguo replied:
The default option is to use server timestamp because that is the timestamp that is printed to the log files.

We provide an option to use the browser timestamp for convenience, but that will not update the timestamp already printed to logs, which will be confusing.

The Ray Dashboard does not parse any logs today. It just prints them as raw text.

alanwguo replied:
I think the simplest approach would be if we used UTC everywhere, but I'm not sure if that's the behavior everyone wants.

A quick fix to the issue at hand would be to drop the timezone names from the UI and only show offset numbers in the dropdown.

aslonnie commented on python/setup.py:
```diff
@@ -264,6 +264,7 @@ def get_packages(self):
             pydantic_dep,
             "prometheus_client >= 0.7.1",
             "smart_open",
+            "tzlocal >= 5.3",  # Required for Ray dashboard timezone handling
```
and I am not a fan of this new dependency..

bhmiller replied:
Can you say more on why? It's already a dependency of yours.

aslonnie replied:
it is not ray dependency, it is a dependency of anyscale CLI, which is currently required to be built into the ray release image to run ray release tests, which is why we added in the dependency resolving procedure and pinning the version for CI/CD. ray does not depend on `tzlocal`. if you `pip install ray[all]` on an empty python virtual env today, `tzlocal` will not be installed.

bhmiller replied:
Ok, if you don't want to use [tzlocal](https://pypi.org/project/tzlocal/) other options I see:
- Drop server time (your previous question). Note in terms of use cases, a remote worker in a different time zone might like using server time zone (though I'm sure that remote worker experiences issues related to this in 200 other places besides the ray dashboard).
- Shortcut it and just use `os.environ["TZ"]` (which skips the full resolution of going to `/etc/timezone` that `tzlocal` does, as I don't want to duplicate what an off the shelf thing does).
- Try and mirror what the typescript side does - e.g. have a "preferred" list of IANA tiime zones to deal with the lack of 1:1 mapping in python:

bhmiller committed:
[Dashboard] Fix timezone_utils to deal with daylight savings

The `Dashboard Server Timezone` is currently incorrect in when
running in a timezone with daylight savings. Fixing that
resolution with python `tzlocal`.

Signed-off-by: Brendan Miller <6900229+bhmiller@users.noreply.github.com>

bhmiller commented:
Rebased this on top of my merged PR in #52755 . So now all that's left in this change is the python changes to fix server time zone. I backed out my attempts to get testing references happy, so expect this to mostly just fail, but I'm otherwise not working on the PR given the feedback/no consensus on how ray folks want to solve this bug (didn't like tzlocal).
